Doc/reference/datamodel.rst

```diff
@@ -804,16 +809,31 @@ Built-in functions
 A built-in function object is a wrapper around a C function.  Examples of
 built-in functions are :func:`len` and :func:`math.sin` (:mod:`math` is a
 standard built-in module). The number and type of the arguments are
-determined by the C function. Special read-only attributes:
+determined by the C function.
 
-* :attr:`!__doc__` is the function's documentation string, or ``None`` if
-  unavailable. See :attr:`function.__doc__`.
-* :attr:`!__name__` is the function's name. See :attr:`function.__name__`.
-* :attr:`!__self__` is set to ``None`` (but see the next item).
-* :attr:`!__module__` is the name of
-  the module the function was defined in or ``None`` if unavailable.
-  See :attr:`function.__module__`.
+Special read-only attributes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment:
Same here, I'm not sure it really makes sense to make this a distinct subsection when there aren't any other subsections inside the datamodel section on builtin functions

Reply:
I thought about that, and I decided by subsection because:

1. consistency with other sections (e.g., the one you highlighted above and Code Objects). In this case, I would fix some other sections, too.
2. personally, I like that (although it is a unique section).
3. if I see the summary, I immediately know the element has or does not have special attributes. More explicit.

Actually, when I opened the issue, I thought about a unique table for both attributes with a column to identify the kind, as the `User-defined functions` section was before your recent changes.

Reply:
Those are all reasonable points! I don't mind the new subsection in the section on builtin functions too much; I'm happy for that to stay.

I _do_ think the new subsection in the section on instance methods makes it more confusing, however. It implies that the entirety of the section on instance methods is about their special read-only attributes, except for the first paragraph. This isn't true: the final paragraph in section 3.2.8.2 has nothing to do with the special read-only methods of instance methods:

> Note that the transformation from [function object](https://cpython-previews--113574.org.readthedocs.build/en/113574/reference/datamodel.html#user-defined-funcs) to instance method object happens each time the attribute is retrieved from the instance. In some cases, a fruitful optimization is to assign the attribute to a local variable and call that local variable. Also notice that this transformation only happens for user-defined functions; other callable objects (and all non-callable objects) are retrieved without transformation. It is also important to note that user-defined functions which are attributes of a class instance are not converted to bound methods; this only happens when the function is an attribute of the class.

Reply:
What if do we move this paragraph to the second paragraph? This seems an explanation for the first one.

Reply:
Well, what about these two paragraphs?

> When an instance method object is created by retrieving a [classmethod](https://cpython-previews--113574.org.readthedocs.build/en/113574/library/functions.html#classmethod) object from a class or instance, its [`__self__`](https://cpython-previews--113574.org.readthedocs.build/en/113574/reference/datamodel.html#method.__self__) attribute is the class itself, and its [`__func__`](https://cpython-previews--113574.org.readthedocs.build/en/113574/reference/datamodel.html#method.__func__) attribute is the function object underlying the class method.
> 
> When an instance method object is called, the underlying function ([`__func__`](https://cpython-previews--113574.org.readthedocs.build/en/113574/reference/datamodel.html#method.__func__)) is called, inserting the class instance ([`__self__`](https://cpython-previews--113574.org.readthedocs.build/en/113574/reference/datamodel.html#method.__self__)) in front of the argument list. For instance, when C is a class which contains a definition for a function f(), and x is an instance of C, calling x.f(1) is equivalent to calling C.f(x, 1).

The first paragraph is about the special read-only attributes of method objects; the second one isn't (it's about a broader topic to do with the semantics of Python methods). So maybe we should move the second paragraph out of the section about the special read-only attributes -- but that would make no sense; it's much more logical to talk about how the `self` reference is inserted at the beginning of the argument list after we've talked about the special `__self__` attribute. The two paragraphs belong together.

For the section on instance methods, I think the existing structure -- where the special read-only attributes are discussed _at the same time_ as other details on method semantics -- makes most sense. Unless you want to propose reworking the whole section on instance methods (which should go into a separate PR, IMO), I don't think we should add a subsection just for the read-only attributes.

Reply:
`it's about a broader topic to do with the semantics of Python methods`

Despite this, I think that we can consider all these paragraphs as about the attributes. They talk about how the attributes work together to element (object, class, method..) to work as well. The attributes are cited a lot of time. 

In first look, I understand this paragraph under attributes this way. However, I also noted some paragraphs didn't are about the attribute, but about the element. I would propose moving these paragraphs to the introduction of the section, probrably. 

Another two idea are:

1. Adding a new section something like "More details", "How it works" about extra content.
2. Removing *all* attribute's subsection mark in the chapter

I prefer first option (although more work), but the second one also provide consistence.

`Unless you want to propose reworking the whole section on instance methods (which should go into a separate PR, IMO),`

Shall we do this together. Some details about it can scape me. :) I think what I told above is in this direction.

(Sorry for any confused thought. It is a brainstorm)